build/Mlos.Cpp.UnitTest.cmake

```diff
@@ -18,3 +18,89 @@ if(NOT googletest_POPULATED)
+    FIXTURES_SETUP MlosSharedMemoriesChecks)
+#set_tests_properties(RemoveMlosSharedMemories PROPERTIES
+#    FIXTURES_CLEANUP MlosSharedMemoriesChecks)
+# Now other test can add "MlosSharedMemoriesChecks" to their
+# "FIXTURES_REQUIRED" property to invoke these scripts before/after themselves.
+
+# Mark all setup/tear down activities involving the shared memory regions as
+# mutually exclusive (no parallel runs).
+set_tests_properties(
+    CheckForMlosSharedMemories
+    #RemoveMlosSharedMemories
+    PROPERTIES RESOURCE_LOCK MlosSharedMemories)
+
+# add_mlos_agent_server_exe_test_run()
+#
+# Provide a function for handling some of the boilerplate to setup a test run
+# of an executable using the Mlos.Agent.Server
+#
+#
+#
+function(add_mlos_agent_server_exe_test_run)
+    set(oneValueArgs NAME EXECUTABLE_TARGET TIMEOUT)
+    set(multiValueArgs MLOS_SETTINGS_REGISTRY_TARGETS)
+    cmake_parse_arguments(add_mlos_agent_server_exe_test_run "${options}" "${oneValueArgs}" "${multiValueArgs}" ${ARGN})
+
+    set(TEST_NAME ${add_mlos_agent_server_exe_test_run_NAME})
+    set(TEST_EXECUTABLE_TARGET ${add_mlos_agent_server_exe_test_run_EXECUTABLE_TARGET})
+
+    if(NOT DEFINED add_mlos_agent_server_exe_test_run_TIMEOUT)
+        set(TEST_TIMEOUT ${DEFAULT_CTEST_TIMEOUT})
+    else()
+        set(TEST_TIMEOUT ${add_mlos_agent_server_exe_test_run_TIMEOUT})
+    endif()
+
+    if(NOT DEFINED add_mlos_agent_server_exe_test_run_MLOS_SETTINGS_REGISTRY_TARGETS)
+        message(SEND_ERROR
+            "Missing MLOS_SETTINGS_REGISTRY_TARGETS for ${TEST_NAME}")
+    endif()
+
+    foreach(SETTINGS_REGISTRY_TGT in ${add_mlos_agent_server_exe_test_run_MLOS_SETTINGS_REGISTRY_TARGETS})
+        if(NOT DEFINED MLOS_SETTINGS_REGISTRY)
+            set(MLOS_SETTINGS_REGISTRY_PATH $<TARGET_PROPERTY:${SETTINGS_REGISTRY_TGT},DOTNET_OUTPUT_DIR>)
+        else()
+            set(MLOS_SETTINGS_REGISTRY_PATH ${MLOS_SETTINGS_REGISTRY_PATH}:$<TARGET_PROPERTY:${SETTINGS_REGISTRY_TGT},DOTNET_OUTPUT_DIR>)
+        endif()
+    endforeach()
+
+    # Basically we want to run:
+    # $ dotnet Mlos.Agent.Server.dll /some/test/exe
+    # However, we need to
+    # - Make sure there aren't other things using the shared mem regions in
+    #   /dev/shm/ that we're about to create (e.g. from a previous failed test
+    #   or something else that's running).
+    #   (and also clean them up after we're done)
+    # - Make sure that the Agent can find the SettingsRegistry DLLs that the
+    #   /some/test/exe needs to use.
+    add_test(NAME ${TEST_NAME}
+        COMMAND ${MLOS_ROOT}/build/CMakeHelpers/RunTestsAndSharedMemChecks.sh ${DOTNET} $<TARGET_PROPERTY:Mlos.Agent.Server,DOTNET_OUTPUT_DLL> $<TARGET_FILE:${TEST_EXECUTABLE_TARGET}>)
+    set_tests_properties(${TEST_NAME} PROPERTIES
+        TIMEOUT ${TEST_TIMEOUT}
+        # Let the Mlos.Agent.Server know where to find the registry assembly.
+        ENVIRONMENT MLOS_SETTINGS_REGISTRY_PATH=${MLOS_SETTINGS_REGISTRY_PATH}
```

Comment:
Another thing we should probably add to avoid this setup overhead for the notebooks is to copy some of the outputs to the `target/bin/{Debug,Release}` directory.